Added CDF and corrected PDF for GumbelDistribution

czgdp1807 opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References
[1] Fixes https://github.com/sympy/sympy/issues/16521
[2] https://en.wikipedia.org/wiki/Gumbel_distribution

#### Brief description of what is fixed or changed
PDF for GumbelDistribution has been corrected with changes made at appropriate places(docs, tests, code)
CDF for GumbelDistribution has been added. Associated test have also been added.


#### Other comments
Code:
```
from sympy import *
import sympy.stats as ss
beta, mu = symbols('beta, mu', real=True)
x = symbols('x', real=True)
X = ss.Gumbel('X', beta, mu)
print(ss.cdf(X).doit())
```
Outputs:
Prior to this addition:
`Lambda(_z, exp(b/eta)*Integral(exp(-_x/eta)*exp(exp(-_x/eta)*exp(b/eta)), (_x, -oo, _z))/eta)`

With this addition:
`Lambda(_z, exp(-exp(_z - b)/eta))`
#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* stats
  * PDF of Gumbel distribution corrected.
  * CDF of Gumbel distribution added.
<!-- END RELEASE NOTES -->

Upabjojr committed:
WIP

Upabjojr committed:
fixing code for matrix derivatives of non-MatrixExpr expressions

asmeurer committed:
Handle matrices in Add and Mul as a post processor

Previously matrices were absorbed into the coeff, but this led to several
bugs, including some that had wrong results.

There is a slight regression here, as seen from the XFAIL test in test_subs.
Namely, substituting multiple matrices into a scalar Add expression will raise a
TypeError because it tries computes the subexpressions. For instance, (x +
y).subs({x: Matrix(...), y: Matrix(...)}). This can be worked around by using
the simultaneous=True flag to subs.

Fixes #15528.

asmeurer committed:
Keep Add(scalar, matrix) unevaluated

scalar + matrix still raises a TypeError.

This allows algorithms to substitute matrices with noncommutative dummy
symbols and manipulate them. subs in particular substitutes each term
iteratively (unless simultaneous=True), so something like (x + y).subs({x:
matrix, y: matrix}) would raise a TypeError without this change. See also the
previous commit message.

asmeurer committed:
Correctly handle the case of no matrices in the MatrixExpr postprocessors

For some reason items get cached in the post-processor and it causes them to
be called on Adds and Muls that don't have any top-level MatrixExprs.

asmeurer committed:
Remove the ability for constructor postprocessors to work with subexpressions

This worked by storing objects in the
Basic._constructor_postprocessor_mapping. However, they were never removed
from that mapping, meaning it could be a memory leak. Furthermore, the fact
that they were there forever meant that some expressions would get
postprocessors called only after other expressions were created, resulting in
nondeterministic undefined behavior.

asmeurer committed:
Remove confusing comment

Now that the postprocesors don't handle subexpressions, the comment isn't
valid. However, I think it is worth leaving the logic there, just in case it
somehow gets triggered still.

asmeurer committed:
Remove explicit check for MatrixExpr in Mul.flatten

They already get added to nc_part anyway.

asmeurer committed:
Make MatMul.args_cnc return a list instead of a tuple

This matches Mul.args_cnc, and is needed for gcd_terms().

asmeurer committed:
Fix gcd_terms with matrix expressions

Mul._from_args can only be used if you already know Mul has processed the
args. If it is an arbitrary expression, or if you didn't check if that
expression is actually a Mul (as opposed to a MatMul), it cannot be used.

asmeurer committed:
Merge branch 'master' into matrices-postprocessor

asmeurer committed:
Call the postprocessors even when evaluate=False

asmeurer committed:
Fix gcd_terms to return a MatMul on matrix expressions

This requires using Mul(evaluate=False) instead of Mul._make_args, so that the
postprocessors get called.

asmeurer committed:
Fix MatMul.args_cnc() tests

asmeurer committed:
Keep noncommutative non-Matrix arguments in order in MatMul.args_cnc()

asmeurer committed:
Set Trace.is_commutative = True

asmeurer committed:
Disallow noncommutative scalars in MatMul

The problem is that some algorithms try to replace matrices with
noncommutative Dummys. But this only works if it can keep them in the right
order, which presently it doesn't, as it separates out the scalar and matrix
parts. Since the current behavior can lead to wrong results, e.g., with
factor_nc(), it is better to explicitly disallow it.

This breaks factor_nc() from working on matrix expressions. However, it only
worked correctly in some cases. For example,
`factor_nc(B**(-1)*(A**(-1)*B**(-1) - A**(-1)*C*B**(-1))**(-1)*A**(-1))`
returned `B**(-1)*(((1 - C)*B**(-1))*A**(-1))**(-1)*A**(-1)`, which not only
reordered the terms incorrectly, but used 1 instead of I.

asmeurer committed:
Better error message

asmeurer committed:
Add an XFAIL test for a matrix expression that could give a wrong result

That is, it would give a wrong result if noncommutative scalars were allowed
in MatMul.

asmeurer committed:
Add a test for diff(Trace(A)*A, A)

See #15121.

asmeurer committed:
Merge branch 'master' into matrices-postprocessor

asmeurer committed:
Merge branch 'master' into matrices-postprocessor

asmeurer committed:
Remove an unused import

smichr committed:
typo

supreet11agrawal committed:
Stats: Added check on the parameters of F Distribution

F Distribution takes two parameters as input which are both positive and
integers. Added checks for this in `crv_types.py` and relevant tests in
`test_continuous_rv.py` of stats module

supreet11agrawal committed:
Increased readability of code

asmeurer committed:
Handle sets that aren't in the expected order in Intersection._handle_finite_sets

The order is usually right because Intersection.__new__ sorts the args first,
but it isn't always. The wrong order leads to an infinite recursion.

Also properly handle the case where Intersection returns UniversalSet.

asmeurer committed:
Raise ValueError when containment of an element cannot be determined when iterating an intersection

This is similar to calling bool() on an undetermined relational (although in
that case we get TypeError instead of ValueError, I'm not sure which is better
here).

See #10249.

smichr committed:
SparseMatrix: accept values that are matrix-like

asmeurer committed:
Clean up the Union and Intersection constructors

- Union and Intersection now accept Python sets as parameters.
- Union and Intersection no longer automatically denest their arguments. If
  there is one argument and is a set, it is returned (union/intersection of a
  single set), otherwise, an exception is raised. This accounts for the
  majority of the changes in this commit.
- Use isinstance in some sets _eval_Eq methods
- Union and Intersection now subclass from LatticeOp. This means among other
  things that they automatically remove duplicate and identity entries.

There is an XFAIL test here which is due to
https://github.com/sympy/sympy/issues/16342, which surfaced some subtle
changes in argument ordering. The test in question now produces a wrong
result, but one could almost certainly produce a similar wrong result before
this commit with different inputs.

asmeurer committed:
Fix some test failures

asmeurer committed:
Add a test that Union and Intersection don't accept iterable single arguments

Also that they do accept Python set arguments, and convert them into
FiniteSets.

sylee957 committed:
Use argparse

sylee957 committed:
Correct formatting

sylee957 committed:
PEP-8 character limit, and also make the log helpful

sylee957 committed:
Format prog in usage

smichr committed:
allow SparseMatrix to autosize

asmeurer committed:
Fix syntax in Python 2

Corwinpro committed:
Update bsplines.py

There is a bug with bspline basis functions construction for a case when the number of control points is small. The removed bit in _add_splines does something weird, while the first part works fine itself.

Here's an example of a wrong behaviour:

import sympy
from sympy.functions.special.bsplines import bspline_basis
x = sympy.var("x")
kv = [0, 0, 0, 0, 1, 2, 2, 2, 2]
basis = bspline_basis(3, kv, 2, x)

# The expected basis function should be continuous at x == 1, while in the current version it is not.
# I computed the expected analytical formula by hand and compared to the one the old version produces. They do not match.

assert basis = sympy.Piecewise((-x**3 + 3*x**2/2, (x >= 0) & (x <= 1)), (x**3 - 9*x**2/2 + 6*x - 2, (x >= 1) & (x <= 2)), (0, True))

With this change we fix the wrong behaviour of the middle basis function, which is only reproducible when the number of control points is small.

Corwinpro committed:
Merge pull request #1 from Corwinpro/Corwinpro-bspline-basis-patch

Update bsplines.py

asmeurer committed:
Revert "Raise ValueError when containment of an element cannot be determined when iterating an intersection"

This reverts commit ff65f86f9da9d848805d944c6d02f9d41c791ace.

Commit:
Add tests to polygon.py

Add tests for the red lines
Improve layout

Commit:
improve layout for line 470

Commit:
quality test passed

Commit:
test, doctest, quality test passed

smichr committed:
detect collisions

smichr committed:
test collision

Commit:
fix error with is_similar

pragyanmehrotra committed:
Added expand() method for SeqFormula

pragyanmehrotra committed:
Added expand() method for SeqFormula

pragyanmehrotra committed:
remove trailing space

pragyanmehrotra committed:
remove trailing spaces

pragyanmehrotra committed:
expand returns a SeqFormula object

oscargus committed:
Fixed some LGTM issues and unmarked slow test to increase coverage

pragyanmehrotra committed:
added tests for expand method for SeqFormula

Commit:
add-perfectgroup-function

asmeurer committed:
Use more compact expression in solve_poly_inequalities

asmeurer committed:
Fix Interval and FiniteSet _eval_Eq logic

dnabanita7 committed:
Changed mpf(0) in evalf_sum to None

As discussed in #11151,S.Zero returned mpf(0) and Sum(0,(a,1,2)) gave
untraceable results of mpf objects.This has to be same as S(0) which gives
none.

smichr committed:
inline edit

smichr committed:
detect singleton lists with ints

smichr committed:
allow control of Matrix.diag from dense.diag

dense.diag is the public facing diag routine that
never got merged with common.diag (accessible through Matrix.diag)
and each routine has different semantics when it comes to
unpacking and treatment of arguments that are lists.

Matrix.diag unpacked singleton sequences and did not convert other
lists to matrices

dense.diag converted all sequences to matrices and hence did not
unpack args

The behavior of dense.diag -- the publicly facing routine imported
with 'from sympy import *' is unchanged.

The behavior of common.diag a.k.a Matrix.diag has been modified to
treat lists as matrices rather than to try place list-literals
into a matrix (which creates a non-functional matrix):

>>> Matrix.diag([1,2],[3])
Matrix([
[[1, 2],   0],
[     0, [3]]])

The "wandering diagonal" behavior is retained so a
block-diagonal can be created from non-square matrices.
If the elements always start filling on the main diagonal,
blocks will end up overlapping. So the folling is correct:

>>> diag([1,2,3],4)
Matrix([
[1, 0],
[2, 0],
[3, 0],
[0, 4]])

This is not a block-diagonal since the "4" block overlaps
with the first column.

>>> diag([1,2,3],4)
Matrix([
[1, 0],
[2, 4],
[3, 0]])

pragyanmehrotra committed:
Remove spaces around = in optional parameter for expand method in SeqFormula

pragyanmehrotra committed:
use *args **kwargs in expand method of SeqFormula

czgdp1807 committed:
Added CDF for GumbelDistribution

CDF for GumbelDistribution has been
added. Associated test have also
been added.

czgdp1807 committed:
Added CDF for GumbelDistribution

CDF for GumbelDistribution has been
added. Associated test have also
been added.

czgdp1807 commented:
@jksuom A trivial addition. Please see if it works.

czgdp1807 committed:
corrected a typo

czgdp1807 committed:
corrected a typo

pragyanmehrotra committed:
fix tests

smichr committed:
Merge pull request #16376 from oscargus/lgtmfixes3

Fixed some LGTM issues and unmarked slow test to increase coverage

smichr committed:
Merge pull request #16344 from asmeurer/union-intersection-no-flatten

Clean up the Union and Intersection constructors

sylee957 committed:
Delete coding utf8

smichr committed:
Merge pull request #15948 from asmeurer/matrices-postprocessor

Handle matrices in Add and Mul as a postprocessor

smichr committed:
Merge pull request #16363 from kangzhiq/16318_add_test

Added tests for polygon.py

asmeurer committed:
Bump the version to 1.4rc1

asmeurer committed:
Bump the version to 1.5.dev

asmeurer committed:
Add a step to release process to create the new release notes page on the wiki

asmeurer committed:
Merge pull request #16389 from asmeurer/release-bump

Bump the version to 1.5.dev

dnabanita7 committed:
Added tests for issue #11151

Upabjojr committed:
fixing bugs

Upabjojr committed:
merge master

Upabjojr committed:
test fix

Upabjojr committed:
Merge pull request #15723 from Upabjojr/trace_derivative_correction

Trace derivative correction

Upabjojr committed:
bug fix on matrix derivative with trace

pragyanmehrotra committed:
add trig = False test case

oscarbenjamin committed:
Finish nth_order_reducible

Upabjojr committed:
Fix derivative of vector in matrices/expressions

Upabjojr committed:
Merge pull request #16397 from Upabjojr/trace_derivative_correction

bug fix on matrix derivative with trace

oscarbenjamin committed:
Fix doctest for ode_nth_order_reducible

sylee957 committed:
Merge pull request #16353 from sylee957/use_argparse_generate_logos

Use argparse for generate_logos.py

oscarbenjamin committed:
Fix type in test_nth_order_reducible

Upabjojr committed:
Merge pull request #16401 from Upabjojr/derive_vectors

Fix derivative of vector in matrices/expressions

dnabanita7 committed:
Added tests for evalf_sum

sidhantnagpal commented on sympy/stats/tests/test_continuous_rv.py:
```diff
@@ -430,6 +430,7 @@ def test_gumbel():
     x = Symbol("x")
     X = Gumbel("x", beta, mu)
     assert simplify(density(X)(x)) == exp((beta*exp((mu - x)/beta) + mu - x)/beta)/beta
+    assert cdf(X)(x) == exp(-exp(-mu + x)/beta)
```
This should return `exp(-exp((mu - x)/beta))`.

czgdp1807 replied:
I have made the corrections. Please check @sidhantnagpal .

oscarbenjamin committed:
Fix underline length in docs

Upabjojr committed:
fixing matrix derivative bugs

wate123 committed:
fixed literal_float regex

dnabanita7 committed:
Added tests

dnabanita7 committed:
Added tests which ae failing and giving exceptions

czgdp1807 committed:
corrected a typo

czgdp1807 committed:
corrected a typo

czgdp1807 committed:
corrected a typo

czgdp1807 committed:
corrected a typo

Commit:
Omit calculation of derived_series if order is odd

oscargus committed:
Fixed issue introduced in #15898

Commit:
update

oscarbenjamin committed:
Fix a few tests for running under pytest

Corwinpro committed:
Update bsplines.py

Fixed doctest for incorrect analytical bspline expression

Commit:
update

smichr committed:
Update common.py

smichr committed:
UndefFunction -> AppliedUndef via _diff_wrt

If an object can act as a differentiation symbol then it is Symbol-like.

smichr committed:
Update piecewise.py

smichr committed:
Update test_piecewise.py

Commit:
Change BaseScalar syntax

Corwinpro committed:
Update bsplines.py

added memoization to bspline_basis (which is called recursively and should be reused)

Corwinpro committed:
Update bsplines.py

removed whitespace and added code formatting with black

smichr committed:
Update piecewise.py

smichr committed:
Update test_pretty.py

smichr committed:
Update test_piecewise.py

smichr committed:
Merge pull request #16422 from kangzhiq/16323_latex_printing_issue

Fixed latex printing of BaseScalar

Commit:
add-test

smichr committed:
Merge pull request #16414 from oscargus/piecewisefix

Fixed issue introduced in #15898

Commit:
add-sylow-test

oscargus committed:
Merge pull request #16368 from pragyanmehrotra/12634_expand_method_for_SeqFormula

Added expand() method for SeqFormula

jmig5776 committed:
Fixed logcombine

smichr committed:
Update test_solvers.py

Commit:
Merge pull request #16410 from oscarbenjamin/pytest_fixes

Fix warnings and warning handling code under pytest

sylee957 committed:
Merge pull request #16386 from sylee957/delete_coding_utf8

Delete unnecessary coding=utf8

jksuom committed:
Merge pull request #16413 from divyanshu132/reduce-is-solvable

Omit calculation of derived_series() if Order is odd

jksuom committed:
Merge pull request #16312 from supreet11agrawal/f_distr_checks

Stats: Added check on the parameters of F Distribution

smichr committed:
Update test_solvers.py

smichr committed:
Merge pull request #16399 from oscarbenjamin/dsolve_reducible

Finish nth_order_reducible

smichr committed:
omit exponential matching in _literal_float

The Float routine will emit an error if the number is not in the right format. The trailing part hasn't been working because of the eE not being written as [eE]. It has been sufficient to see that the number starts like a floating point number so that is not being changed.

To be more strict and only match an exact float -- including the exponential part -- the following tested diff could be applied:

diff --git a/sympy/core/numbers.py b/sympy/core/numbers.py
index b7c01bd..735dcc4 100644
--- a/sympy/core/numbers.py
+++ b/sympy/core/numbers.py
@@ -147,10 +147,10 @@ def _decimal_to_Rational_prec(dec):
     return rv, prec


-def _literal_float(f):
+_refloat = regex.compile(r"[-+]?((\d*\.\d+)|(\d+\.?\d?))([eE][-+]?\d+)?")
+def _literal_float(f, d=False):
     """Return True if n can be interpreted as a floating point number."""
-    pat = r"[-+]?((\d*\.\d+)|(\d+\.?))(eE[-+]?\d+)?"
-    return bool(regex.match(pat, f))
+    return bool(_refloat.fullmatch(f))

 # (a,b) -> gcd(a,b)
 
bad = []
good=[]
for p in cartes('+- ','1 ','. ','1 ','eE ','+- ','1 '):
 n = ''.join(p).replace(' ','')
 try:
   ok=float(n)
   good.append(n)
 except: bad.append(n)
assert len(good) == 152
assert len(bad) == 280
from sympy.core.numbers import _refloat
assert all(_refloat.fullmatch(i) for i in good)
assert not any(_refloat.fullmatch(i) for i in bad)

Commit:
update-test

smichr committed:
Update test_solvers.py

shivanikohlii committed:
underscore added

smichr committed:
Merge pull request #16409 from wate123/issue16407

fixed incorrect literal_float regex

dnabanita7 committed:
Added tests for #11151

smichr committed:
Update test_solvers.py

supreet11agrawal reviewed (changes requested):
https://github.com/sympy/sympy/blob/b475b451b3ad89270baa33a5939a0035a5c5343e/sympy/stats/crv_types.py#L1486
Below this, perhaps `cdf` example can be added.

sidhantnagpal commented on sympy/stats/tests/test_continuous_rv.py:
```diff
@@ -430,6 +430,7 @@ def test_gumbel():
     x = Symbol("x")
     X = Gumbel("x", beta, mu)
     assert simplify(density(X)(x)) == exp((beta*exp((mu - x)/beta) + mu - x)/beta)/beta
+    assert cdf(X)(x) == exp(-exp((-mu + x)/beta))
```
The sign is still not correct. It should be `mu - x`, see the formula at https://en.wikipedia.org/wiki/Gumbel_distribution#Properties.

czgdp1807 replied:
@sidhantnagpal Corrected. Apologies for a lot of mistakes.

ritesh99rakesh committed:
added tests for particle and rigidbody in physics/mechanics

dnabanita7 committed:
Added tests to #11151

dnabanita7 committed:
Added 3 separate tests for #11151

ritesh99rakesh committed:
added tests for functions under physics/mechanics

shivanikohlii committed:
Added test for underscore workaround

asmeurer committed:
Update AUTHORS and .mailmap

Upabjojr committed:
added support for derivatives of ElementwiseApplyFunction

smichr committed:
pep8 and condense into one test

smichr committed:
Update test_evalf.py

smichr committed:
Merge pull request #16429 from jmig5776/fix_logcombine

combine single unknown log with others

Upabjojr committed:
Hadamard power added

Upabjojr committed:
added HadamardPower to test_args

Corwinpro committed:
Update bsplines.py

Fixed caching of list instead of tuple kv

smichr committed:
Update test_evalf.py

asmeurer committed:
Install gcc in the release Docker using apt

smichr committed:
Merge pull request #16402 from Naba7/mpf_evalf

Changed mpf(0) to None in evalf_sum

smichr committed:
add diagonal to common

ritesh99rakesh committed:
typo error corrected

Upabjojr committed:
fixing tests and restoring Transpose in core

czgdp1807 committed:
corrected sign

czgdp1807 committed:
corrected sign

czgdp1807 committed:
updated tests

czgdp1807 committed:
updated tests

Corwinpro committed:
Update bsplines.py

Fixed list -> tuple in bspline basis set

Upabjojr committed:
added str printer for HadamardPower

ritesh99rakesh committed:
removed assertion statment

Upabjojr committed:
remove extra return

Corwinpro committed:
Update bsplines.py

Back compatibility with Python2 added: range() is not hashable in python2.7

moorepants committed:
Merge pull request #16440 from ritesh99rakesh/16318

added tests for particle and rigidbody in physics/mechanics

smichr committed:
Merge pull request #16343 from smichr/sparse

SparseMatrix: accept values that are matrix-like

smichr committed:
Merge pull request #16341 from smichr/diag2

merge common.diag and dense.diag

jksuom committed:
Merge pull request #16359 from Corwinpro/master

Fix Bspline _add_bsplines

smichr committed:
add __mod__ to matrix

smichr committed:
add note about finding gcd of matrix

ShubhamKJha committed:
Possible fix for Posify

Upabjojr committed:
adding some tests

oscarbenjamin committed:
Use ignore_warnings instead of warns for numpy

shivanikohlii committed:
changes as suggested by smichr

smichr committed:
Update numbers.py

smichr committed:
Update test_numbers.py

jksuom committed:
Merge pull request #16375 from divyanshu132/PerfectGroup

Added PerfectGroup Function

smichr committed:
Update numbers.py

oscarbenjamin committed:
Merge pull request #16450 from ShubhamKJha/posify_fix_2

Ensure that posify preserves unrelated assumptions such as finite, integer etc.

asmeurer committed:
Merge pull request #16455 from oscarbenjamin/fix_numpy_warning

Use ignore_warnings instead of warns for numpy

asmeurer committed:
Add a note about rebuilding the Docker container

smichr committed:
diagonal modifications

oscargus committed:
Fixed issues with default values

oscargus committed:
Fixed issues in ode.py

oscargus committed:
Fixed potential integer division issues

oscargus committed:
Fixed redundant checks and missing ,

oscargus committed:
Removed unused imports and fixed typo

asmeurer committed:
Matplotlib is now required to build the docs

oscargus committed:
Removed unused variables

oscargus committed:
Removed self-imports

oscargus committed:
Changed some duplicated loop variables

oscargus committed:
Removed unused imports

smichr committed:
Merge pull request #16437 from shivanikohlii/Issue-13638

Fixes float misplace decimal issue

asmeurer committed:
Update git whitelist for the release scripts

asmeurer committed:
Merge branch '1.4' of github.com:sympy/sympy into 1.4

asmeurer committed:
The release tarball is a .tar.gz, not just .tar

asmeurer committed:
Fix tarball test in the release script

asmeurer committed:
Merge branch '1.4' of github.com:sympy/sympy into 1.4

Upabjojr committed:
Merge pull request #16443 from Upabjojr/hadamard_power

Hadamard power added

Upabjojr committed:
Merge branch 'master' into matrix_applyfunc_derivative

smichr committed:
test invalid diagonal references

oscargus committed:
Fixed review comments

smichr committed:
Merge pull request #16454 from smichr/block

Matrix __mod__ and diagonal defined

hackerde committed:
Change return type of sort_key function

The return type for this function defined in `partitions.py` was list.
The error being raised was that the type list was not hashable.
But tuple is hashable, so according to the suggestions, I changed the return value into a tuple.
I wrote a test in `test_partitions.py` and it passed when I ran it with `bin/test`.

czgdp1807 commented:
@sidhantnagpal Any updates/comments on the PR?

sidhantnagpal commented on sympy/stats/crv_types.py:
```diff
@@ -1442,6 +1442,10 @@ def pdf(self, x):
         beta, mu = self.beta, self.mu
         return (1/beta)*exp(-((x-mu)/beta)+exp(-((x-mu)/beta)))
 
+    def cdf(self, x):
```
The method name `_cdf` is used for the probability distributions.
```suggestion
    def _cdf(self, x):
```

Commit:
used _cdf instead of cdf

this is done for maintaining consistency and depicting private scope of cdf.

Co-Authored-By: czgdp1807 <singh.23@iitj.ac.in>

Commit:
used _cdf instead of cdf

this is done for maintaining consistency and depicting private scope of cdf.

Co-Authored-By: czgdp1807 <singh.23@iitj.ac.in>

czgdp1807 commented:
@sidhantnagpal Please let me know if there is anything else to change.

oscargus committed:
Imported CompileError directly from distutils.errors

oscargus committed:
Used or-style

smichr committed:
add See Also entries to diag*

Upabjojr committed:
Merge pull request #16442 from Upabjojr/matrix_applyfunc_derivative

added support for derivatives of ElementwiseApplyFunction

smichr committed:
Merge pull request #16482 from hackerde/9608_partitions

Change return type of sort_key function

Upabjojr committed:
Merge pull request #16408 from Upabjojr/matrix_deriv_1.4

fixing matrix derivative bugs

smichr committed:
Update exprtools.py

smichr committed:
Merge pull request #16463 from oscargus/lgtmfixes4

Code quality fixes

sylee957 committed:
Correct links

The links are supposed to work like this

smichr committed:
Merge pull request #16487 from smichr/diag

add See Also entries to diag*

oscargus committed:
Added some tests to physics to increase coverage

asmeurer committed:
Use SHA 256 instead of MD5 in the release script

asmeurer committed:
Remove Python 3.4 support

Python 3.4 has reached its [end of
life](https://devguide.python.org/#status-of-python-branches), so by our
[Python version support
policy](https://github.com/sympy/sympy/wiki/Python-version-support-policy) we
are no longer supporting it.

SymPy 1.4 will be the last version to support Python 3.4. Now that the 1.4
release branch has been made, we no longer need to support it on master.

It is unlikely this will buy us anything in the way of Python features, since
we still are supporting Python 2.7. However, it does mean fewer tests on
Travis.

smichr committed:
Merge pull request #16491 from oscargus/physicstests

Added some tests to physics to increase coverage

smichr committed:
Merge pull request #16492 from asmeurer/rm-34

Remove Python 3.4 support

sidhantnagpal commented:
Looks fine to me, can you add an example for the docs?

czgdp1807 committed:
Added docs example

czgdp1807 committed:
Added docs example

czgdp1807 committed:
added missing import in docs

czgdp1807 committed:
added missing import in docs

sidhantnagpal commented:
It is also worth noting that with appropriate assumptions set, the result currently is the following (which also has an `oo` term):
```
In [2]: mu = Symbol("mu")                                                       

In [3]: beta = Symbol("beta", positive=True)                                    

In [4]: X = Gumbel("x", beta, mu)

In [5]: cdf(X)(x)                                                              
Out[5]: 
⎛           μ  -x     ⎞   
⎜     -μ    ─  ───    ⎟  μ
⎜     ───   β   β     ⎟  ─
⎜      β   ℯ ⋅ℯ       ⎟  β
⎝- β⋅ℯ   ⋅ℯ        + ∞⎠⋅ℯ 
──────────────────────────
            β             
```

The problem is that the formula for density (pdf) of Gumbel is wrong.
It seems it was introduced in https://github.com/sympy/sympy/pull/11751.

With the correct formula for density(X)(x), cdf(X)(x) does give the correct result, however it might be slow, so the changes here (#16381) are OK.

czgdp1807 committed:
corrected pdf of gumbel distribution

czgdp1807 committed:
corrected pdf of gumbel distribution

czgdp1807 committed:
Merge branch 'stats_simple' of https://github.com/czgdp1807/sympy into stats_simple

czgdp1807 committed:
updated test

czgdp1807 committed:
updated test

czgdp1807 commented:
@sidhantnagpal Kindly check, the updates. 
I believe that using pre-existing results is useful since, it helps in speeding up the process of generating output. Moreover, the correctness is not affected irrespective of the changes made in modules outside sympy.stats.

sidhantnagpal commented on sympy/stats/crv_types.py:
```diff
@@ -1440,7 +1440,12 @@ class GumbelDistribution(SingleContinuousDistribution):
         beta, mu = self.beta, self.mu
-        return (1/beta)*exp(-((x-mu)/beta)+exp(-((x-mu)/beta)))
+        z = (x-mu)/beta
+        return (1/beta)*exp(-(z+exp(-z)))
```
Maybe just `exp(-(z + exp(-z)))/beta`. (Consider adding spaces around binary operators `+` and `-`.)

czgdp1807 replied:
I forgot to do that here. Thanks for correcting.

sidhantnagpal commented on sympy/stats/tests/test_continuous_rv.py:
```diff
@@ -429,7 +429,8 @@ def test_gumbel():
     x = Symbol("x")
     X = Gumbel("x", beta, mu)
-    assert simplify(density(X)(x)) == exp((beta*exp((mu - x)/beta) + mu - x)/beta)/beta
+    assert simplify(density(X)(x)) == exp(-exp(-(-mu + x)/beta) - (-mu + x)/beta)/beta
```
Can we drop the `simplify` here?

czgdp1807 replied:
Agreed, it just slows down the code. This shouldn't have been there. I don't know the reason placing it there.

czgdp1807 replied:
I got it. It shouldn't be dropped. May be, there are comparison issues. Let's keep it there.

sidhantnagpal replied:
The tests could be just:
```diff
--- a/sympy/stats/tests/test_continuous_rv.py
+++ b/sympy/stats/tests/test_continuous_rv.py
@@ -443,7 +443,8 @@ def test_gumbel():
     mu = Symbol("mu")
     x = Symbol("x")
     X = Gumbel("x", beta, mu)
-    assert simplify(density(X)(x)) == exp((beta*exp((mu - x)/beta) + mu - x)/beta)/beta
+    assert str(density(X)(x)) == 'exp(-exp(-(-mu + x)/beta) - (-mu + x)/beta)/beta'
+    assert cdf(X)(x) == exp(-exp((mu - x)/beta))
```

czgdp1807 committed:
Update crv_types.py

czgdp1807 committed:
Update crv_types.py

czgdp1807 committed:
Update test_continuous_rv.py

czgdp1807 committed:
Update test_continuous_rv.py

sidhantnagpal commented:
Let's help the docstring of `Gumbel`, as well. The following should do it:
```diff
@@ -1462,15 +1463,16 @@ def Gumbel(name, beta, mu):
     The density of the Gumbel distribution is given by
 
     .. math::
-        f(x) := \exp \left( -exp \left( x + \exp \left( -x \right) \right) \right)
+        f(x) := \dfrac{1}{\beta} \exp \left( -\dfrac{x-\mu}{\beta}
+                - \exp \left( -\dfrac{x - \mu}{\beta} \right) \right)
 
-    with ::math 'x \in [ - \inf, \inf ]'.
+    with :math:`x \in [ - \infty, \infty ]`.
```

czgdp1807 committed:
updated docstrings

czgdp1807 committed:
updated docstrings

czgdp1807 commented:
@sidhantnagpal Done. Please see if it LGTY.

sidhantnagpal commented:
Looking good. Can you rebase it onto 1.4 branch and squash the commits with an appropriate message?

czgdp1807 committed:
Merge branch '1.4' of git://github.com/sympy/sympy into 1.4

czgdp1807 committed:
Update test_continuous_rv.py

czgdp1807 committed:
Update test_continuous_rv.py

czgdp1807 committed:
Merge branch 'stats_simple' of https://github.com/czgdp1807/sympy into stats_simple

czgdp1807 committed:
updated tests

czgdp1807 committed:
updated tests

czgdp1807 committed:
updated tests

czgdp1807 committed:
Merge branch 'stats_simple' of https://github.com/czgdp1807/sympy into stats_simple
